#Pipeline to analyze Cyanobacteria genomes from the raw fastqc reads to quality assessment and feature annotation.

**Usage:**

1. Clone this repository to your machine;
2. Add the raw bacterial Illumina reads to a subdirectory called *rawReads* insinde the directory. (note: Illumina pair reads
  should have the standard output names ending in **R1_001.fastq** and **R2_001.fastq**);
3. Use `make` to run the data analysis or `make clean` to remove the outputs of your previous analysis;
4. ???
5. Profit!
